tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionEvaluator.cs

```diff
@@ -0,0 +1,233 @@
+                {
+                    HandleException(ref result, compiledExpressions[i], e.Message);
+                }
+            }
+
+            var finalMessage = StringBuilderCache.GetStringAndRelease(resultBuilder);
+            result.Template = finalMessage;
+        }
+        catch (Exception e)
+        {
+            result.Template = e.Message;
+            (result.Errors ??= new List<EvaluationError>()).Add(new EvaluationError { Expression = null, Message = e.Message });
+        }
+    }
+
+    private void EvaluateCondition(ref ExpressionEvaluationResult result)
+    {
+        CompiledExpression<bool> compiledExpression = default;
+        try
+        {
+            if (!CompiledCondition.Value.HasValue)
+            {
+                return;
+            }
+
+            compiledExpression = CompiledCondition.Value.Value;
+            var condition = compiledExpression.Delegate(ScopeMembers.InvocationTarget, ScopeMembers.Members);
+            result.Condition = condition;
+            if (compiledExpression.Errors != null)
+            {
+                (result.Errors ??= new List<EvaluationError>()).AddRange(compiledExpression.Errors);
+            }
+        }
+        catch (Exception e)
+        {
+            HandleException(ref result, compiledExpression, e.Message);
+            result.Condition = true;
+        }
+    }
+
+    private void EvaluateMetric(ref ExpressionEvaluationResult result)
+    {
+        CompiledExpression<double> compiledExpression = default;
+        try
+        {
+            if (!CompiledMetric.Value.HasValue)
+            {
+                return;
+            }
+
+            compiledExpression = CompiledMetric.Value.Value;
+            var metric = compiledExpression.Delegate(ScopeMembers.InvocationTarget, ScopeMembers.Members);
+            result.Metric = metric;
+            if (compiledExpression.Errors != null)
+            {
+                (result.Errors ??= new List<EvaluationError>()).AddRange(compiledExpression.Errors);
+            }
+        }
+        catch (Exception e)
+        {
+            HandleException(ref result, compiledExpression, e.Message);
+        }
+    }
+
+    private CompiledExpression<string>[] CompileTemplates()
+    {
+        var compiledExpressions = new CompiledExpression<string>[Templates.Length];
+        for (int i = 0; i < Templates.Length; i++)
+        {
+            var current = Templates[i];
+            if (current.Json != null)
+            {
+                compiledExpressions[i] = ProbeExpressionParser<string>.ParseExpression(current.Json, ScopeMembers.InvocationTarget, ScopeMembers.Members);
+            }
+            else if (current.Str != null)
+            {
+                compiledExpressions[i] = new CompiledExpression<string>(null, null, current.Str, null);
+            }
+            else
+            {
+                throw new Exception($"{nameof(CompileTemplates)}: Template segment must have json or str");
```

Comment:
Maybe also include the template index?
And do we really want to propagate an exception and not mark the specific expression/template as 'failed' and keep the rest that are properly constructed?